### PR TITLE
fix #28369, transpose of SparseMatrixCSC is not recursive

### LINKED ISSUE #28369: Inconsistency in descending `transpose` between sparse and dense arrays
```julia
using LinearAlgebra
using SparseArrays
A = [1 0 ; 1 0]

julia> M = reshape([A, A, A, A], (2,2))
2×2 Array{Array{Int64,2},2}:
 [1 0; 1 0]  [1 0; 1 0]
 [1 0; 1 0]  [1 0; 1 0]

julia> copy(transpose(M))
2×2 Array{Array{Int64,2},2}:
 [1 1; 0 0]  [1 1; 0 0]
 [1 1; 0 0]  [1 1; 0 0]

julia> Array(copy(transpose(sparse(M))))
2×2 Array{Array{Int64,2},2}:
 [1 0; 1 0]  [1 0; 1 0]
 [1 0; 1 0]  [1 0; 1 0]
```
Ideally, I expect sparse and dense objects to represent the same thing, but to employ different representations internally. I convert between them only to take advantage of differences in algorithmic efficiency.

I guess there is no perfect solution to this kind of thing. Computer-language types and mathematical-object types (sets) are far from congruent. E.g., in math, it is common to interpret objects in one section of a manuscript or book in a particular way without worrying about breaking examples in another section. Even the most flexible computer-language type system is more rigid.

### PR DESCRIPTION
**Pull request by jlapeyre**

### CODE REVIEW

File: stdlib/SparseArrays/test/sparse.jl [hunk 1]
```diff
@@ -2253,4 +2253,22 @@ end
...
+    M = reshape([[1 2; 3 4], [9 10; 11 12], [5 6; 7 8], [13 14; 15 16]], (2,2))
+    MP = reshape([[1 2; 3 4], [5 6; 7 8], [9 10; 11 12], [13 14; 15 16]], (2,2))
+    S = sparse(M)
+    SP = sparse(MP)
+    @test isa(transpose(S), Transpose)
+    @test transpose(S) == copy(transpose(S))
+    @test Array(transpose(S)) == copy(transpose(M))
+    @test permutedims(S) == SP
+    @test permutedims(S, (2,1)) == SP
+    @test permutedims(M, (1,2)) == S
```
Comment (Sacha0) [inline stdlib/SparseArrays/test/sparse.jl]: Did you perhaps mean to test `permutedims(S, (1,2)) == M` instead?

Comment (jlapeyre) [reply stdlib/SparseArrays/test/sparse.jl]: `permutedims(M, (1,2)) == S`  doesn't test the new code. `permutedims(S, (1,2)) == M` is definitely better.

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 2]
```diff
@@ -867,8 +867,14 @@ function ftranspose(A::SparseMatrixCSC{Tv,Ti}, f::Function) where {Tv,Ti}
...
-Base.copy(A::Adjoint{<:Any,<:SparseMatrixCSC}) = ftranspose(A.parent, conj)
-Base.copy(A::Transpose{<:Any,<:SparseMatrixCSC}) = ftranspose(A.parent, identity)
+Base.copy(A::Adjoint{<:Any,<:SparseMatrixCSC}) = ftranspose(A.parent, x -> copy(adjoint(x)))
+Base.copy(A::Transpose{<:Any,<:SparseMatrixCSC}) = ftranspose(A.parent, x -> copy(transpose(x)))
+function Base.permutedims(A::SparseMatrixCSC, (a,b))
+    (a, b) == (2, 1) && return ftranspose(A, identity)
+    (a, b) == (1, 2) && return copy(A)
+    throw(ArgumentError("no valid permutation of dimensions"))
+end
+Base.permutedims(A::SparseMatrixCSC) = permutedims(A, (2,1))
```
Comment (jlapeyre) [inline stdlib/SparseArrays/src/sparsematrix.jl]: This line is redundant. It is covered by the `permutedims(A::AbstractMatrix) = permutedims(A, (2,1)` in *permuteddimsarray.jl*

File: stdlib/SparseArrays/test/sparse.jl (hunk 1 above)
Comment (jlapeyre) [reply stdlib/SparseArrays/test/sparse.jl]: I fixed this and rebased and pushed the PR again.

File: stdlib/SparseArrays/src/sparsematrix.jl (hunk 2 above)
Comment (jlapeyre) [reply stdlib/SparseArrays/src/sparsematrix.jl]: I fixed this and rebased and pushed the PR again

**Comment by Sacha0** [comment]
A nanosoldier run might've been in order, to make certain that the replacement of the function passed to `ftranspose` did not introduce any performance regressions?

**Comment by andreasnoack** [comment]
Another time please just go ahead and trigger a run.

@nanosoldier `runbenchmarks("array", vs = "@c05fd201baedb6681947cc212e0a9f49564d1568")`

**Comment by jlapeyre** [comment]
Is this something I can do locally ?

**Comment by andreasnoack** [comment]
Yes, you can run the same tests locally. See https://github.com/JuliaCI/BaseBenchmarks.jl. Though, it might be easier to ask a collaborator to trigger a nanosoldier run.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/28376#issuecomment-409869255) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/24d217eacfc10123c7204e38360b72df947665ef/56b504a_vs_c05fd20/report.md). cc @ararslan

**Comment by jlapeyre** [comment]
There appear to be significant regressions (+ 30-40%) in some sparse benchmarks. I'll look at this.

**Comment by andreasnoack** [comment]
I'm not convinced that any of these are genuine but it would be great if you could take a closer look anyway

**Comment by jlapeyre** [comment]
I misread the results. Time increases of about %30 are precisely on those with about %30 tolerance.

But, there is an oversight in the PR. 
`transpose!` and `adjoint!` should also be recursive. 
https://github.com/JuliaLang/julia/blob/b0bf91e7cf761d960329e3133dd4c3978e1732a8/stdlib/SparseArrays/src/sparsematrix.jl#L858-L859

**Comment by Sacha0** [comment]
The collection of sparse-dense and dense-sparse matrix multiplication regressions look potentially legitimate; that they occur as a block makes the likelihood that those represent a real change reasonable. Perhaps check those locally?

**Comment by jlapeyre** [comment]
It looks like doing this correctly will take a bit more work. As someone mentioned, the sparsematrix.jl code relies in a few places on the elements being scalars.

Suppose
`S = sparse(reshape([[1 2; 3 4], [9 10; 11 12], [5 6; 7 8], [13 14; 15 16]], (2,2)))`.

 One problem is unnecessary allocation. `copy(transpose(S))` calls `ftranspose`, which allocates a container
https://github.com/JuliaLang/julia/blob/883c8a38920985e0c02df169ff8c379731d88fc6/stdlib/SparseArrays/src/sparsematrix.jl#L861-L865

 Eventually, `_distributevals_halfperm!` is called, which rebinds the top-level elements in the allocated container to the (materially) transposed sub-matrices.

This PR neglected to fix `tranpose!`. If we fix it by again replacing `identity` with `x -> copy(tranpose(x))` here
https://github.com/JuliaLang/julia/blob/883c8a38920985e0c02df169ff8c379731d88fc6/stdlib/SparseArrays/src/sparsematrix.jl#L858
the same additional allocation will happen.

But, I haven't done or tested anything yet because `Revise` is broken for the moment on Julia master.  I guess that redundant allocation is preferable in the short term to silently giving the wrong result.

**Pull request merged**